lahja/endpoint.py

```diff
@@ -145,8 +145,9 @@ def _process_item(self, item: BaseEvent, config: BroadcastConfig) -> None:
 
         if in_futures:
             future = self._futures[config.filter_event_id]
-            future.set_result(item)
-            self._futures.pop(config.filter_event_id)
+            if not future.cancelled() and not future.done():
+                future.set_result(item)
+                self._futures.pop(config.filter_event_id)
```

Comment:
Hrm, wouldn't you still want to do this pop regardless of done/cancelled status?

Reply:
Mmh...I don't think so because if we ever get here when the future is already done it will mean that it probably is removed already either because we hit this exact code location before (from another thread) or because we hit this when the future was cancelled.

https://github.com/ethereum/lahja/blob/5e3cd14815df2c279e67bba5dd78475b46741331/lahja/endpoint.py#L227

But now that I think about it, I guess we should play it safe whenever we attempt to remove a key from `self._futures` and check if the key still exists.

In fact, because of the fact that `_process_item` may be called from multiple threads I'm really wondering if we even can avoid using thread locks here. But I would like to try to avoid those if we can.

Reply:
You can just do a `self._futures.pop(id, None)` which will work gracefully if the key isn't present.